src/Icicle/Benchmark.hs

```diff
@@ -0,0 +1,219 @@
+  fleet <- firstEitherT BenchSeaError (seaCompile cfg avalanche)
+
+  end <- liftIO getCurrentTime
+
+  return Benchmark {
+      benchSource          = code
+    , benchAssembly        = asm
+    , benchFleet           = fleet
+    , benchInputPath       = inputPath
+    , benchOutputPath      = outputPath
+    , benchChordPath       = packedChordPath
+    , benchCompilationTime = end `diffUTCTime` start
+    }
+
+releaseBenchmark :: Benchmark -> EitherT BenchError IO ()
+releaseBenchmark b =
+  seaRelease (benchFleet b)
+
+runBenchmark :: Benchmark -> EitherT BenchError IO BenchmarkResult
+runBenchmark b = do
+  let fleet      = benchFleet      b
+      inputPath  = benchInputPath  b
+      outputPath = benchOutputPath b
+      chordPath  = benchChordPath  b
+
+  start <- liftIO getCurrentTime
+  stats <- firstEitherT BenchSeaError (seaPsvSnapshotFilePath fleet inputPath outputPath chordPath)
+  end   <- liftIO getCurrentTime
+
+  size <- liftIO (withFile inputPath ReadMode hFileSize)
+
+  return BenchmarkResult {
+      benchTime     = end `diffUTCTime` start
+    , benchFacts    = psvFactsRead    stats
+    , benchEntities = psvEntitiesRead stats
+    , benchBytes    = fromIntegral size
+    }
+
+tombstonesOfDictionary :: Dictionary -> Map Attribute (Set Text)
+tombstonesOfDictionary dict =
+  let go (DictionaryEntry a (ConcreteDefinition _ ts)) = [(a, ts)]
+      go _                                             = []
+  in Map.fromList (concatMap go (dictionaryEntries dict))
+
+------------------------------------------------------------------------
+
+avalancheOfDictionary :: Dictionary -> Either BenchError (Map Attribute (A.Program (Annot ()) S.Variable A.Prim))
+avalancheOfDictionary dict = do
+  let virtuals = fmap (second unVirtual) (getVirtualFeatures dict)
+
+  core      <- parTraverse (coreOfSource dict) virtuals
+  fused     <- parTraverse fuseCore (Map.unionsWith (<>) core)
+  avalanche <- parTraverse avalancheOfCore fused
+
+  return avalanche
+
+avalancheOfCore :: C.Program () S.Variable -> Either BenchError (A.Program (Annot ()) S.Variable A.Prim)
+avalancheOfCore core = do
+  flat    <- first BenchAvalancheError (coreFlatten core)
+  checked <- first BenchAvalancheError (checkAvalanche flat)
+  return checked
+
+fuseCore :: [(S.Variable, C.Program () S.Variable)] -> Either BenchError (C.Program () S.Variable)
+fuseCore programs =
+  first BenchFusionError $ do
+    fused <- C.fuseMultiple () programs
+    pure (coreSimp fused)
+
+coreOfSource
+  :: Dictionary
+  -> (Attribute, QueryTop'T SourceVar)
+  -> Either BenchError (Map Attribute [(S.Variable, C.Program () S.Variable)])
+coreOfSource dict (Attribute attr, virtual) =
+  first BenchSourceError $ do
+    let inlined = sourceInline dict virtual
+
+    desugared    <- sourceDesugarQT inlined
+    (checked, _) <- sourceCheckQT dict desugared
+
+    let reified = sourceReifyQT checked
+
+    core <- sourceConvert dict reified
+    let simplified = coreSimp core
```

Comment:
this (and above) is a bit like Pipeline stuff? Is it worth putting them in the main pipeline for repl and tests to use?